File: src/org/irislang/jiris/core/IrisThreadInfo.java
package org.irislang.jiris.core;

import jdk.nashorn.internal.runtime.Debug;
import org.irislang.jiris.dev.IrisDevUtil;
import org.irislang.jiris.irisclass.IrisInteger;
import org.omg.CORBA.PRIVATE_MEMBER;
import sun.misc.Cleaner;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Set;
import java.util.Stack;

public class IrisThreadInfo {
	
	private static long sm_mainThreadID = 0;
	private static IrisThreadInfo sm_mainThreadInfo = null;
	
	public static IrisThreadInfo GetCurrentThreadInfo() {
		long threadID = Thread.currentThread().getId();
		
		if(threadID == sm_mainThreadID) {
			return sm_mainThreadInfo;
		} else {
			return null;
		}
	}
	
	public static void SetMainThreedInfo(IrisThreadInfo threadInfo) {
		sm_mainThreadInfo = threadInfo;
	}
	
	public static void SetMainThreadID(long id) {
		sm_mainThreadID = id;
	}

	private ArrayList<IrisValue> m_parameterList = new ArrayList<IrisValue>();
	private IrisValue m_record = null;

	private IrisValue m_comparedObj = null;
	private int m_counter = 0;
	private Stack<IrisValue> m_loopTimeStack = new Stack<IrisValue>();
	private Stack<IrisValue> m_comparedObjectStack = new Stack<IrisValue>();
	private Stack<IrisValue> m_counterStack = new Stack<IrisValue>();
	private Stack<IrisValue> m_vesselStack = new Stack<IrisValue>();
	private Stack<IrisValue> m_iteratorStack = new Stack<IrisValue>();
	private Stack<IrisContextEnvironment> m_environmentStack = new Stack<IrisContextEnvironment>();

	private IrisValue m_tempSuperClass = null;
	private LinkedList<IrisValue> m_tempModules = new LinkedList<IrisValue>();
	private LinkedList<IrisValue> m_tempInterfaces =  new LinkedList<IrisValue>();

	public void SetTempSuperClass(IrisValue value) {
	    m_tempSuperClass = value;
    }

    public IrisValue GetTempSuperClass() {
	    return m_tempSuperClass;
    }

    public void ClearTempModules() {
        m_tempModules.clear();
    }

    public void AddTempModule(IrisValue value) {
	    m_tempModules.add(value);
    }

    public LinkedList<IrisValue> GetTempModules() {
	    return m_tempModules;
    }

    public void ClearTempInterfaces() {
        m_tempInterfaces.clear();
    }

    public void AddTempInterface(IrisValue value) {
	    m_tempInterfaces.add(value);
    }

    public LinkedList<IrisValue> getTempInterfaces() {
	    return m_tempInterfaces;
    }

	public void PushContext(IrisContextEnvironment environment) {
		m_environmentStack.push(environment);
	}

	public IrisContextEnvironment PopContext() {
		return m_environmentStack.pop();
	}

	public void PushVessel(IrisValue vessel) {
		m_vesselStack.push(vessel);
	}

	public IrisValue GetVessel() {
		return m_vesselStack.lastElement();
	}

	public void PopVessel() {
		m_vesselStack.pop();
	}

	public void PushIterator(IrisValue iterator) {
		m_iteratorStack.push(iterator);
	}

	public IrisValue GetIterator() {
		return m_iteratorStack.lastElement();
	}

	public void PopIterator() {
		m_iteratorStack.pop();
	}

	public void PushComparedObject(IrisValue value) { m_comparedObjectStack.add(value);}

	public void PopCompareadObject() { m_comparedObjectStack.pop(); }

	public IrisValue GetTopComparedObject() { return m_comparedObjectStack.lastElement(); }

	public void PushLoopTime(IrisValue value) {
		m_loopTimeStack.add(value);
	}

	public void PopLoopTime() {
		m_loopTimeStack.pop();
	}

	public IrisValue GetTopLoopTime() {
		return m_loopTimeStack.lastElement();
	}
	
	public IrisValue getCounter() {
		return m_counterStack.lastElement();
	}
	
	public void pushCounter(IrisValue counter) {
		m_counterStack.push(counter);
	}

	public void PopCounter() {
		m_counterStack.pop();
	}

	public void increamCounter() {
		IrisInteger.IrisIntegerTag tag = (IrisInteger.IrisIntegerTag)IrisDevUtil.GetNativeObjectRef(getCounter());
		tag.setInteger(tag.getInteger() + 1);
//		System.out.print("\ncounter : ");
//		System.out.print(tag.getInteger());
//		System.out.print("\n");
	}
	
	public ArrayList<IrisValue> getPartPrameterListOf(int count) {
		return new ArrayList<IrisValue>(m_parameterList.subList(m_parameterList.size() - count, m_parameterList.size()));
	}
	
	public ArrayList<IrisValue> getParameterList() {
		return m_parameterList;
	}
	
	public void AddParameter(IrisValue value) {
		m_parameterList.add(value);
	}
	
	public void PopParameter(int times) {
		for(int i = 0; i < times; ++i) {
			m_parameterList.remove(m_parameterList.size() - 1);
		}
	}
	
	public void ClearPrameterList() {
		m_parameterList.clear();
	}
		
	public IrisValue getRecord() {
		return m_record;
	} 
	
	public void setRecord(IrisValue parameter) {
		m_record = parameter;
	}
}

File: src/main/java/org/irislang/jiris/core/IrisThreadInfo.kt
package org.irislang.jiris.core

import org.irislang.jiris.compiler.statement.IrisClassStatement
import com.sun.deploy.panel.UpdatePanelFactory
import jdk.nashorn.internal.runtime.Debug
import org.irislang.jiris.dev.IrisDevUtil
import org.irislang.jiris.irisclass.IrisInteger
import org.omg.CORBA.PRIVATE_MEMBER
import sun.misc.Cleaner

import javax.print.attribute.standard.MediaSize
import java.lang.invoke.MethodHandle
import java.util.ArrayList
import java.util.LinkedList
import java.util.Stack

class IrisThreadInfo {

    val parameterList = ArrayList<IrisValue>()
    var record: IrisValue? = null

    private val m_comparedObj: IrisValue? = null
    private val m_counter = 0
    private val m_loopTimeStack = Stack<IrisValue>()
    private val m_comparedObjectStack = Stack<IrisValue>()
    private val m_counterStack = Stack<IrisValue>()
    private val m_vesselStack = Stack<IrisValue>()
    private val m_iteratorStack = Stack<IrisValue>()
    private val m_environmentStack = Stack<IrisContextEnvironment>()
    private val m_closureBlockStack = Stack<IrisClosureBlock>()

    private var m_tempSuperClass: IrisValue? = null
    private val m_tempModules = LinkedList<IrisValue>()
    val tempInterfaces = LinkedList<IrisValue>()

    var currentLineNumber = -1
    var currentFileName: String = ""

    val counter: IrisValue
        get() = m_counterStack.lastElement()

    fun GetTopContextEnvironment(): IrisContextEnvironment {
        return if(m_environmentStack.isEmpty()) IrisContextEnvironment()  else m_environmentStack.lastElement()
    }

    fun GetTopClosureBlock(): IrisClosureBlock {
        return m_closureBlockStack.lastElement()
    }

    fun PushClosureBlock(block: IrisClosureBlock) {
        m_closureBlockStack.push(block)
    }

    fun PopClosureBlock() {
        m_closureBlockStack.pop()
    }

    fun SetTempSuperClass(value: IrisValue) {
        m_tempSuperClass = value
    }

    fun GetTempSuperClass(): IrisValue? {
        return m_tempSuperClass
    }

    fun ClearTempModules() {
        m_tempModules.clear()
    }

    fun AddTempModule(value: IrisValue) {
        m_tempModules.add(value)
    }

    fun GetTempModules(): LinkedList<IrisValue> {
        return m_tempModules
    }

    fun ClearTempInterfaces() {
        tempInterfaces.clear()
    }

    fun AddTempInterface(value: IrisValue) {
        tempInterfaces.add(value)
    }

    fun PushContext(environment: IrisContextEnvironment) {
        m_environmentStack.push(environment)
    }

    fun PopContext(): IrisContextEnvironment {
        return m_environmentStack.pop()
    }

    fun PushVessel(vessel: IrisValue) {
        m_vesselStack.push(vessel)
    }

    fun GetVessel(): IrisValue {
        return m_vesselStack.lastElement()
    }

    fun PopVessel() {
        m_vesselStack.pop()
    }

    fun PushIterator(iterator: IrisValue) {
        m_iteratorStack.push(iterator)
    }

    fun GetIterator(): IrisValue {
        return m_iteratorStack.lastElement()
    }

    fun PopIterator() {
        m_iteratorStack.pop()
    }

    fun PushComparedObject(value: IrisValue) {
        m_comparedObjectStack.add(value)
    }

    fun PopCompareadObject() {
        m_comparedObjectStack.pop()
    }

    fun GetTopComparedObject(): IrisValue {
        return m_comparedObjectStack.lastElement()
    }

    fun PushLoopTime(value: IrisValue) {
        m_loopTimeStack.add(value)
    }

    fun PopLoopTime() {
        m_loopTimeStack.pop()
    }

    fun GetTopLoopTime(): IrisValue {
        return m_loopTimeStack.lastElement()
    }

    fun pushCounter(counter: IrisValue) {
        m_counterStack.push(counter)
    }

    fun PopCounter() {
        m_counterStack.pop()
    }

    fun increamCounter() {
        val tag = IrisDevUtil.GetNativeObjectRef<Any>(counter) as IrisInteger.IrisIntegerTag
        tag.integer = tag.integer + 1
    }

    fun getPartPrameterListOf(count: Int): ArrayList<IrisValue> {
        return ArrayList(parameterList.subList(parameterList.size - count, parameterList.size))
    }

    fun AddParameter(value: IrisValue) {
        parameterList.add(value)
    }

    fun PopParameter(times: Int) {
        for (i in 0 until times) {
            parameterList.removeAt(parameterList.size - 1)
        }
    }

    fun ClearPrameterList() {
        parameterList.clear()
    }

    companion object {

        private var sm_mainThreadID: Long = 0
        private var sm_mainThreadInfo: IrisThreadInfo? = null

        fun GetCurrentThreadInfo(): IrisThreadInfo? {
            val threadID = Thread.currentThread().id

            return if (threadID == sm_mainThreadID) {
                sm_mainThreadInfo
            } else {
                null
            }
        }

        fun SetMainThreedInfo(threadInfo: IrisThreadInfo) {
            sm_mainThreadInfo = threadInfo
        }

        fun SetMainThreadID(id: Long) {
            sm_mainThreadID = id
        }
    }
}
